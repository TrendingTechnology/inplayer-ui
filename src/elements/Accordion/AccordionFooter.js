// @flow
import React from 'react';
import styled from 'styled-components';
import colors from 'config/colors';
import { uiColors, fontWeights } from 'utils';
import { ifProp } from 'styled-tools';

import Button from 'elements/Button';

type Props = {
  footerLink: string,
  buttonText: string,
  isOpen: boolean,
  toggleClose: () => boolean,
};

const AccordionFooterContainer = styled.footer`
  position: absolute;
  bottom: ${ifProp('isOpen', '0', '-80px')};
  width: 100%;
  box-sizing: border-box;
  display: flex;
  justify-content: space-between;
  padding: 1rem 3%;
  background: ${colors.white};
  align-items: center;
  transition-timing-function: ease, ease;
  transition-duration: 400ms, 1000ms;
  transition-property: all, background;
  transition-delay: 0s, 300ms;
  z-index: ${ifProp('isOpen', '40', '10')};
  border-top: 1px solid ${colors.gray};
  border-bottom: 1px solid ${colors.gray};
`;

const AccordionFooterLink = styled.a`
  color: ${uiColors('text.light')};
  font-weight: ${fontWeights('light')};
  cursor: pointer;
`;

const AccordionFooter = ({ footerLink, buttonText, toggleClose, isOpen }: Props) => (
  <AccordionFooterContainer isOpen={isOpen}>
    <AccordionFooterLink onClick={toggleClose(false)}>{footerLink}</AccordionFooterLink>
    <Button size="sm" buttonModifiers={['buttonPrimary']} onClick={toggleClose(true)}>
      {buttonText}
    </Button>
  </AccordionFooterContainer>
);

export default AccordionFooter;
