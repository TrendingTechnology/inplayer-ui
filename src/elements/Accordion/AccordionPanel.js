// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import colors from 'config/colors';
import { ifProp } from 'styled-tools';
import { uiColors, fontWeights } from 'utils';

// Components
import Icon from 'elements/Icon';
import Typography from 'elements/Typography';
import Tooltip, { Props as TooltipProps } from 'components/Tooltip';

const AccordionPanelContainer = styled.div`
  width: 100%;
  height: 3.5rem;
  box-sizing: border-box;
  overflow: hidden;
  z-index: 10;
  background: ${colors.white};
  transition: ease 500ms height;
  display: none;
  ${ifProp(
    'isActive',
    css`
      position: absolute;
      ${({ contentHeight }) => contentHeight && `height: calc(${contentHeight} - 15.5rem)`};
      top: 0;
      z-index: 30;
      border-bottom: none;
      overflow: auto;
      margin-top: 3.5rem;
      display: block;
    `
  )};
`;

const AccordionPanelHeader = styled.header`
  margin: 0;
  padding: 1rem 3%;
  width: 100%;
  box-sizing: border-box;
  background: ${ifProp('disabled', colors.lightGray, colors.white)};
  cursor: ${({ disabled, isActive }) => (disabled || isActive ? 'default' : 'pointer')};
  border-bottom: 1px solid ${colors.lightGray};
  display: flex;
  justify-content: space-between;
  align-items: center;
  ${ifProp(
    'isActive',
    css`
      position: absolute;
      top: 0;
    `
  )};
`;

const AccordionTitle = styled(Typography)`
  font-weight: ${fontWeights('thin')};
  color: ${ifProp('disabled', uiColors('text.disabled'), uiColors('text.main'))};
  margin: 0;
  display: inline;
`;

const AccordionIcon = styled(Icon)`
  color: ${uiColors('primary.main')};
  ${({ pointer }) => (pointer ? 'cursor: pointer;' : '')}
`;

const AccordionPanelDetails = styled.div`
  padding: 2rem 6%;
`;

type Props = {
  label: string,
  isActive: boolean,
  isOtherPanelActive: boolean,
  icon?: string,
  iconTooltip?: TooltipProps,
  contentHeight: string,
  openPanel: (name: string) => boolean,
  renderContent: () => any,
  disabled: boolean,
};

const AccordionPanel = ({
  openPanel,
  label,
  isActive,
  isOtherPanelActive,
  icon,
  iconTooltip,
  renderContent,
  contentHeight,
  disabled,
}: Props) => (
  <>
    <AccordionPanelHeader
      disabled={disabled}
      onClick={!disabled ? openPanel : null}
      isActive={isActive}
    >
      <AccordionTitle variant="h6" isActive={isActive} disabled={disabled}>
        {label}
      </AccordionTitle>
      {!isOtherPanelActive &&
        (iconTooltip ? (
          <Tooltip {...iconTooltip}>
            <AccordionIcon name={icon} pointer />
          </Tooltip>
        ) : (
          <AccordionIcon name={icon} />
        ))}
    </AccordionPanelHeader>
    <AccordionPanelContainer isActive={isActive} contentHeight={contentHeight}>
      <AccordionPanelDetails>{isActive && renderContent()}</AccordionPanelDetails>
    </AccordionPanelContainer>
  </>
);

AccordionPanel.defaultProps = {
  icon: '',
  iconTooltip: null,
};

export default AccordionPanel;
