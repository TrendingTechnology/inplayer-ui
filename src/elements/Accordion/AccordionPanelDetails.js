// @flow
import React from 'react';
import styled from 'styled-components';

const AccordionPanelDetailsContainer = styled.div`
  padding: 10px 6% 26px;
`;

const AccordionPanelDetils = () => <AccordionPanelDetailsContainer />;

export default AccordionPanelDetils;
