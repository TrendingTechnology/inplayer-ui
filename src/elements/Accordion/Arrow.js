// @flow
import React from 'react';
import styled from 'styled-components';
import { ifProp } from 'styled-tools';
import colors from 'config/colors';
import { transparentize } from 'polished';

export const ArrowContainer = styled.div`
  width: 18px;
  margin: 0 1rem 1rem 1rem;
  align-self: flex-start;
  text-align: center;

  &:hover i {
    border-color: ${colors.gray};
  }
`;

export const NavbarArrow = styled.i`
  border: solid ${transparentize(0.6, colors.blue)};
  border-width: 1px 0 0 1px;
  display: inline-block;
  padding: 0.225rem;
  transform: ${ifProp('open', 'rotate(135deg)', 'rotate(-45deg)')};
  transition: all 0.5s ease;
  cursor: pointer;
`;

type Props = {
  open: boolean,
};

const Arrow = ({ open }: Props) => (
  <ArrowContainer>
    <NavbarArrow open={open} />
  </ArrowContainer>
);

export default Arrow;
