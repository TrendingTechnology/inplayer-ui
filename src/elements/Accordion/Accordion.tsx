import React, { useState, SyntheticEvent } from 'react';

// Components
import { Props as TooltipProps } from 'components/Tooltip';
import { AccordionWrapper } from './styled';
import AccordionPanel from './AccordionPanel';
import Arrow from '../../components/NavBar/Arrow';

// Types
type Panel = {
  label: string;
  icon?: string;
  iconTooltip?: TooltipProps;
  renderContent: () => any;
  disabled: boolean;
};

type Props = {
  panels: Array<Panel>;
  contentHeight: string;
  width?: string;
  extendWidth?: string;
  isExtendable?: boolean;
  onActivePanelChange?: (index: number) => void;
};

const Accordion = ({
  panels,
  contentHeight,
  width,
  extendWidth,
  isExtendable,
  onActivePanelChange,
}: Props) => {
  const [activePanel, setActivePanel] = useState(-1);
  const [open, setOpen] = useState(false);

  const openPanel = (panelIndex: number) => {
    if (panelIndex !== activePanel) {
      setActivePanel(panelIndex);
      if (onActivePanelChange) onActivePanelChange(panelIndex);
    }
  };

  const closePanel = (e?: SyntheticEvent) => {
    if (e) e.stopPropagation();
    setActivePanel(-1);
  };

  const togglePanel = (panelIndex: number) => (e: SyntheticEvent<any, Event>) => {
    e.stopPropagation();
    if (activePanel !== -1) {
      closePanel();
    } else {
      openPanel(panelIndex);
    }
  };

  const toggleOpen = () => {
    if (isExtendable) setOpen(!open);
  };

  return (
    <AccordionWrapper
      width={width}
      extendWidth={extendWidth}
      open={open}
      contentHeight={contentHeight}
    >
      <div>
        {panels.map((panel, index) => {
          const { icon, iconTooltip, label, renderContent, disabled } = panel;
          const isActive = activePanel === index;
          const isOtherPanelActive = !isActive && activePanel !== -1;

          return (
            <div key={index}>
              <AccordionPanel
                key={index}
                isActive={isActive}
                icon={icon}
                iconTooltip={iconTooltip}
                label={label}
                renderContent={renderContent}
                closePanel={closePanel}
                isOtherPanelActive={isOtherPanelActive}
                contentHeight={contentHeight}
                disabled={disabled}
                togglePanel={togglePanel(index)}
              />
            </div>
          );
        })}
      </div>
      {isExtendable && activePanel === -1 && (
        <Arrow onClick={toggleOpen} open={open} section="acordion" />
      )}
    </AccordionWrapper>
  );
};

export default Accordion;
