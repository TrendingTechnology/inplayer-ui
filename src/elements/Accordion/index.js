// @flow
import React, { Component, type Node } from 'react';

// Components
import { AccordionWrapper } from './styled';
import AccordionPanel from './AccordionPanel';
import AccordionFooter from './AccordionFooter';
import Arrow from '../../components/NavBar/Arrow';

type Panel = {
  label: string,
  icon?: string,
  renderContent: () => any,
  renderActionButton: ({
    closeAccordion: (e?: SyntheticEvent<*>) => void,
  }) => Node,
};

type Props = {
  panels: Array<Panel>,
  contentHeight: string,
  width?: string,
  extendWidth?: string,
  enableExtendAccordion?: boolean,
};

type State = { activePanel: number, open: boolean };

class Accordion extends Component<Props, State> {
  state = { activePanel: -1, open: false };

  openPanel = (panelIndex: number) => (e: SyntheticEvent<*>) => {
    e.stopPropagation();
    const { activePanel } = this.state;

    if (panelIndex !== activePanel) {
      this.setState({ activePanel: panelIndex });
    }
  };

  closePanel = (e?: SyntheticEvent<*>) => {
    if (e) e.stopPropagation();
    this.setState({ activePanel: -1 });
  };

  toggleOpen = () => {
    const { enableExtendAccordion } = this.props;
    console.log('HERERE');
    if (enableExtendAccordion) this.setState(({ open }) => ({ open: !open }));
  };

  render() {
    const { panels, contentHeight, width, extendWidth, enableExtendAccordion } = this.props;

    const { state } = this;
    const { open } = state;

    return (
      <AccordionWrapper
        width={width}
        extendWidth={extendWidth}
        open={open}
        contentHeight={contentHeight}
      >
        <div>
          {panels.map((panel, index) => {
            const { icon, label, renderActionButton, renderContent } = panel;
            const { activePanel } = state;

            return (
              <div key={index}>
                <AccordionPanel
                  key={index}
                  isActive={activePanel === index}
                  icon={icon}
                  label={label}
                  renderContent={renderContent}
                  openPanel={this.openPanel(index)}
                  closePanel={this.closePanel}
                  contentHeight={contentHeight}
                />
                <AccordionFooter
                  isActive={activePanel === index}
                  closePanel={this.closePanel}
                  renderActionButton={renderActionButton}
                />
              </div>
            );
          })}
        </div>
        {enableExtendAccordion && (
          <Arrow onClick={this.toggleOpen} open={open} section="accordion" />
        )}
      </AccordionWrapper>
    );
  }
}

Accordion.defaultProps = {
  width: '100%',
  extendWidth: '20%',
  enableExtendAccordion: false,
};

/** @component */
export default Accordion;
