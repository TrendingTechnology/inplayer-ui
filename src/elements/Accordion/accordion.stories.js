// @flow
import React from 'react';
import styled from 'styled-components';
import { uiColors } from 'utils';

// components
import { MdInfo } from 'react-icons/md';
import Accordion from './index';
import Typography from '../Typography/index';

export default {
  component: Accordion,
  title: 'Accordion',
};

const Container = styled.div`
  margin: auto;
  margin-top: 2%;
  width: 90%;
`;

const StyledIcon = styled(MdInfo)`
  color: ${uiColors('primary.main')};
`;

export const demo = () => {
  const panels = [
    {
      label: 'Accordion1',
      icon: <StyledIcon />,
      renderContent: ({ closePanel }) => (
        <div>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
          <Typography variant="h1" color="black">
            Long Content for the accordion1
          </Typography>
        </div>
      ),
    },
    {
      label: 'Accordion2',
      icon: <StyledIcon />,
      iconTooltip: {
        content: 'This is a disabled panel with optional icon with a tooltip',
        placement: 'left',
      },
      renderContent: ({ closePanel }) => <h5>Content for the accordionh2</h5>,
      disabled: true,
    },
    {
      label: 'Accordion3',
      renderContent: () => (
        <Typography variant="h5" color="black">
          Content for the accordion h3
        </Typography>
      ),
    },
  ];

  return (
    <Container>
      <Typography variant="h1" color="black">
        Accordion
      </Typography>
      <Typography variant="p" description>
        Accordion or expansion panel is a component that generates Accordion component with more
        accordion panels.It takes <b>panels</b> - array of objects and each object should contain:
        <br /> - <b>label</b> (title)
        <br /> - <b>icon</b> (icon displayed next to the title)
        <br /> - <b>renderContent</b> (action used to render the content)
        <br /> - <b>iconTooltip</b> (optional parameter - a tooltip to display on hovering the icon,
        see <b>Tooltip</b>'s prop list) - default <b>null</b>
        <br /> - <b>disabled</b>(optional paremeter - a boolean to disable clicking or opening the
        panel) - default <b>false</b>
        <br />
        <br />
        Other props that the accordion accepts:
        <br /> - <b>isExtendable</b> (display arrow and allow the accordion to extend)
        <br /> - <b>width</b> (optional parameter - sets the width for the Accordion wrapper) -
        default 100 %
        <br /> - <b>extendWidth</b> (optional parameter - defines the width the accordion should
        extend) - default 20 %
        <br /> - <b>contentHeight</b> - that determines the height of the wrapper.
        <br /> - <b>onActivePanelChange</b> - function to be executed when the accordion tab changes
      </Typography>
      <Accordion
        onActivePanelChange={(index) => console.log(index)}
        panels={panels}
        contentHeight="100vh"
        isExtendable
        width="80%"
        extendWidth="20%"
      />
    </Container>
  );
};
