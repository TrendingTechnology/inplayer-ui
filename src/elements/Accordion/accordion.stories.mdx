import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Accordion from '../Accordion/'
import { panels } from './panels.tsx'

<Meta title="Accordion" component={<Accordion/>} />

# Accordion

<Props of={Accordion}/>

Accordion or expansion panel is a component that generates Accordion component with more accordion panels. It takes panels - array of objects and each object should contain:
- label (title)
- icon (icon displayed next to the title)
- renderContent (action used to render the content)
- iconTooltip (optional parameter - a tooltip to display on hovering the icon, see Tooltip's prop list) - default null
- disabled (optional paremeter - a boolean to disable clicking or opening the panel) - default false

Other props that the accordion accepts:
- isExtendable (display arrow and allow the accordion to extend)
- width (optional parameter - sets the width for the Accordion wrapper) - default 100%
- extendWidth (optional parameter - defines the width the accordion should extend) - default 20%
- contentHeight - that determines the height of the wrapper.
- onActivePanelChange - function to be executed when the accordion tab changes

```jsx
import { Accordion } from '@inplayer-org/inplayer-ui';

const Page = () => (
    <Accordion
        onActivePanelChange={(index) => console.log(index)}
        panels={panels}
        contentHeight="46vh"
        isExtendable
        width="80%"
        extendWidth="20%"
    />
)
```

### Panels example

```jsx
import { MdInfo } from 'react-icons/md';

const panels = [
  {
    label: 'Accordion1',
    icon: <MdInfo color="orange" />,
    renderContent: () => (
      <div>
        <h1>Long Content for the accordion1</h1>
        <h1>Long Content for the accordion1</h1>
        <h1>Long Content for the accordion1</h1>
        <h1>Long Content for the accordion1</h1>
        <h1>Long Content for the accordion1</h1>
        <h1>Long Content for the accordion1</h1>
      </div>
    ),
  },
  {
    label: 'Disabled Accordion',
    icon: <MdInfo color="gray" />,
    renderContent: () => (
      <div>
        <h1>Long Content for the accordion</h1>
        <h1>Long Content for the accordion</h1>
        <h1>Long Content for the accordion</h1>
        <h1>Long Content for the accordion</h1>
        <h1>Long Content for the accordion</h1>
        <h1>Long Content for the accordion</h1>
      </div>
    ),
    disabled: true,
  },
  {
    label: 'Accordion3',
    icon: <MdInfo color="green" />,
    renderContent: () => (
      <div>
        <h1>Long Content for the accordion3</h1>
        <h1>Long Content for the accordion3</h1>
        <h1>Long Content for the accordion3</h1>
        <h1>Long Content for the accordion3</h1>
        <h1>Long Content for the accordion3</h1>
        <h1>Long Content for the accordion3</h1>
      </div>
    ),
  },
];

```

### 
<Preview withSource="open">
  <Story name="accordion">
    <div>
        <Accordion
            onActivePanelChange={(index) => console.log(index)}
            panels={panels}
            contentHeight="46vh"
            isExtendable
            width="80%"
            extendWidth="20%"
        />
    </div>
  </Story>
</Preview>

