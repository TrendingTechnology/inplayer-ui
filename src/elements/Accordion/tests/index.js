import { generateSnapshots } from 'helpers';

import Accordion from '../index';

generateSnapshots(Accordion, {
  panels: [
    {
      label: 'Accordion',
      icon: '',
      content: 'Lorem Ipsum',
      saveAction: () => {
        console.log('Do something!');
      },
    },
  ],
});
