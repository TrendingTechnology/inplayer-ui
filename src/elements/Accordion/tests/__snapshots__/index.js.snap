// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly 1`] = `
.c0 {
  width: 100%;
  position: relative;
  background: #fff;
  border-left: 1px solid #d7dde5;
  box-sizing: border-box;
  border-top: 1px solid #d7dde5;
  overflow-y: hidden;
}

.c4 {
  font-family: Roboto;
  font-weight: 100;
  color: #286090;
  font-size: 1.25rem;
}

.c1 {
  position: relative;
  width: 100%;
  height: 3.5rem;
  box-sizing: border-box;
  overflow: hidden;
  z-index: 10;
  background: #fff;
  -webkit-transition: ease 500ms height;
  transition: ease 500ms height;
}

.c2 {
  margin: 0;
  padding: 0.75rem 3%;
  width: 100%;
  box-sizing: border-box;
  background: #fff;
  border-bottom: 1px solid #F6F8FB;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  font-weight: 100;
  color: #44494e;
  margin: 0;
  cursor: pointer;
  display: inline;
}

.c5 {
  display: inline-block;
  font-size: 1rem;
  color: #07AAE6;
}

.c6 {
  padding: 0.625rem 6% 26px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <header
      className="c2"
      onClick={[Function]}
    >
      <h6
        className="c3 c4"
        color=""
        style={Object {}}
      >
        Accordion
      </h6>
      <span
        aria-hidden={true}
        className="fa fa- c5"
      />
    </header>
    <div
      className="c6"
    >
      <div>
        Lorem Ipsum
      </div>
    </div>
  </div>
</div>
`;
